### PR TITLE
Fallback custom fields

### PR DESCRIPTION
**Pull request by cabloo**

### CODE REVIEW

File: src/Server/Provision/ServerProvisioner.php [hunk 1]
```diff
@@ -234,12 +234,19 @@ private function getIps()
...
+        $disks = $this->multiChoice($choices, '/Drive Bay ([0-9]+)(.*)/');
+        $addons = $this->addons($this->multiChoice($choices, '/Add On ([0-9]+)/'));
+
+        $preconfiguredMemory = $this->getConfigValues(WhmcsConfig::MEM_BILLING_ID, 'Memory', false, true);
+        $preconfiguredDisks = $this->getConfigValues(WhmcsConfig::DISK_BILLING_IDS, 'Drive Bay', false, false);
+        $preconfiguredAddons = $this->addons($this->getConfigValues(WhmcsConfig::ADDON_BILLING_IDS, 'Add On', false, false));
 
         return [
-            'mem_billing' => $this->config->getOption('Memory'),
+            'mem_billing' => $preconfiguredMemory ?: $memory,
```
Comment (cabloo) [inline src/Server/Provision/ServerProvisioner.php]: this is supposed to be the other way around, if there is a configurable option, it should override the preconfigured selections.

File: src/Server/Provision/ServerProvisioner.php [hunk 2]
```diff
@@ -361,6 +386,59 @@ private function createTicket(array $params)
         ]);
     }
 
+    /**
+     * @param string $configID
+     * @param string $newKey
+     * @param bool $getNames    tells the function to return the frontend names that users see
```
Comment (cabloo) [inline src/Server/Provision/ServerProvisioner.php]: Boolean arguments are [usually bad](https://medium.com/@amlcurran/clean-code-the-curse-of-a-boolean-parameter-c237a830b7a3), as with this case. Better to have separate functions usually and have those functions share a common function call inside them for the parts that would be duplicated.

File: src/Server/Provision/ServerProvisioner.php [hunk 3]
```diff
@@ -361,6 +386,59 @@ private function createTicket(array $params)
...
+            return $this->parseConfigOptions($configNamesAndValues[0], $newKey, $singular);
+        }
+
+        return $this->parseConfigOptions($configValue, $newKey, $singular);
+    }
+
+    /**
+     * @param string $configOptions
+     * @param string $newKey
+     * @param bool $singular    tells the function not to append a number to the end of the billing name if it a field that is named in a singular way.
```
Comment (cabloo) [inline src/Server/Provision/ServerProvisioner.php]: see above

File: src/Server/Provision/ServerProvisioner.php [hunk 4]
```diff
@@ -393,8 +471,6 @@ private function getRaidLevel()
      */
     private function addons(array $choices)
     {
-        $addons = $this->multiChoice($choices, '/Add On ([0-9]+)/');
```
Comment (cabloo) [inline src/Server/Provision/ServerProvisioner.php]: why was this removed? Looks like it's needed below?

File: src/Whmcs/WhmcsConfig.php [hunk 5]
```diff
@@ -15,13 +15,16 @@ class WhmcsConfig
...
      */
     const CPU_BILLING_ID = 1;
-    const API_USER = 2;
-    const TICKET_DEPT = 3;
-    const PXE_ACCESS = 4;
-    const IPMI_ACCESS = 5;
-    const SWITCH_ACCESS = 6;
-    const DELETE_ACTION = 7;
-    const PRE_INSTALL = 8;
+    const MEM_BILLING_ID = 2;
```
Comment (cabloo) [inline src/Whmcs/WhmcsConfig.php]: Fairly certain it is not safe to change the values here, since now the old value that was in API_USER would now be stored in MEM_BILLING_ID (the 2 is how it's stored in the DB).

File: src/Whmcs/WhmcsConfig.php [hunk 6]
```diff
@@ -30,6 +33,9 @@ class WhmcsConfig
      */
     protected $countOptions = self::PRE_INSTALL;
 
+    const MEM_BILLING_DESC = 'Billing ID of the RAM. Setting this will override user-selected RAM options. ex: mem-1 | 8 GB RAM';
```
Comment (cabloo) [inline src/Whmcs/WhmcsConfig.php]: text will need to be updated per above comments about precedence. This should also include something to the effect of "this field does nothing when the Memory Configurable Option is present".

File: synergycp.php [hunk 7]
```diff
@@ -4,7 +4,7 @@
     die('This file cannot be accessed directly.');
 }
 
-ini_set('display_errors', 'Off');
+ini_set('display_errors', 'On');
```
Comment (cabloo) [inline synergycp.php]: shouldn't change this on the production version

File: src/Server/Provision/ServerProvisioner.php (hunk 4 above)
Comment (Natehooper) [reply src/Server/Provision/ServerProvisioner.php]: I moved it outside of the function because I need to call the function and dont want that run on my code, there's a very similar function for disks and its called outside of that one but not this one.

Comment (Natehooper) [reply src/Server/Provision/ServerProvisioner.php]: its a private function and is only called once in the code so i figure its nbd

File: synergycp.php (hunk 7 above)
Comment (Natehooper) [reply synergycp.php]: i just have this set in mine, i'll remove it before the final PR

File: src/Server/Provision/ServerProvisioner.php (hunk 1 above)
Comment (Natehooper) [reply src/Server/Provision/ServerProvisioner.php]: ah i see, what threw me off was "Our system already has CPU configured this way but we need a way for WHMCS admins to be able to configure all the hardware on a per-product basis without the user being able to override their choice." so it sounded like the user dropdown choice would be moot at that point, but I see that they just won't have the dropdowns if thats the case.

Comment (cabloo) [reply src/Server/Provision/ServerProvisioner.php]: That was meant to be "_with_ the user being able to override their choice."

File: src/Server/Provision/ServerProvisioner.php (hunk 4 above)
Comment (cabloo) [reply src/Server/Provision/ServerProvisioner.php]: My point is, on the next line, there's an `array_filter($addons)`. That `$addons` variable no longer exists since it was removed here.

File: src/Server/Provision/ServerProvisioner.php [hunk 8]
```diff
@@ -238,11 +238,16 @@ public function getFilters()
         $disks = $this->multiChoice($choices, '/Drive Bay ([0-9]+)(.*)/');
         $addons = $this->addons($this->multiChoice($choices, '/Add On ([0-9]+)/'));
 
+        $configAddons = $this->getConfigValues(WhmcsConfig::ADDON_BILLING_IDS);
+        if ($configAddons) {
```
Comment (cabloo) [inline src/Server/Provision/ServerProvisioner.php]: try to avoid updating the contents of a variable as it's confusing - especially if the type of data stored in the variable changes.

File: clientarea.tpl [hunk 9]
```diff
@@ -118,19 +118,22 @@
...
-    <hr />
+  <hr />
+  <form method="post" action="{$url_action}">
+    <input type="hidden" name="a" value="btn_manage" />
+    <input type="submit" value="Manage on SynergyCP" class="btn btn-info" />
+  </form>
+  <br />
+{/if}
 
+{if $embed}
```
Comment (cabloo) [inline clientarea.tpl]: this shouldn't be part of this PR?

File: README.md [hunk 10]
```diff
@@ -53,3 +53,7 @@ Every SynergyCP product on WHMCS must include the following Configurable Options
 | Drive Bay 1, Drive Bay 2, etc. (any number of entries) | 1TB-HDD&#124;1 TB Hard Drive                              | The billing ID value should be None for empty disk bays.                                                     |
 | Operating System                                       | os-centos6&#124;CentOS 6                                  | The billing ID must match an OS Reload Profile's Billing ID on SynergyCP                                     |
 | Bandwidth                                              | 20000GB&#124;20 TB Bandwidth or &#124;Unmetered Bandwidth | Optional. The billing ID **MUST** use the same unit as your overage billing setting in WHMCS (default is GB) |
+
+**Preset Configurable Options**:
+
+Under Products/Services < Module Settings there are the MEM Billing ID, Addon Billing IDs, and Disk Billing IDs fields. These can be set as default preset values for these fields. If the corresponding Configurable Option field is included it will override its respective field. These fields use a comma-separated format that allows for multiple values: add-1, add-2|Addon1, Addon2
```
Comment (cabloo) [inline README.md]: can you mention the intended use of this, i.e. you can have products that are completely or partially pre-configured for companies that offer a pre-defined set of hardware configurations?

File: src/Server/Provision/ServerProvisioner.php [hunk 11]
```diff
@@ -259,13 +259,22 @@ private function getIps()
     public function getFilters()
     {
         $choices = $this->config->options();
+        $memory = $this->config->getOption('Memory');
+        $disks = $this->multiChoice($choices, '/Drive Bay ([0-9]+)(.*)/');
+        $addons = $this->addons($this->multiChoice($choices, '/Add On ([0-9]+)/'));
+        $configAddons = $this->addons($this->getConfigBillingValues(WhmcsConfig::ADDON_BILLING_IDS));
+
+        // $configAddons = $this->getConfigBillingValues(WhmcsConfig::ADDON_BILLING_IDS);
```
Comment (cabloo) [inline src/Server/Provision/ServerProvisioner.php]: please delete commented code

File: src/Server/Provision/ServerProvisioner.php [hunk 12]
```diff
@@ -370,7 +379,26 @@ private function createTicket(array $params)
             $params['password']
         );
 
+        $presetConfigOptions = array_filter(array_merge(
+            [
+                'Memory' => $this->getConfigName(WhmcsConfig::MEM_BILLING_ID),
+            ],
+            // $this->getConfigName(WhmcsConfig::MEM_BILLING_ID, 'Memory') ?: [],
```
Comment (cabloo) [inline src/Server/Provision/ServerProvisioner.php]: delete commented code

File: src/Whmcs/WhmcsConfig.php [hunk 13]
```diff
@@ -22,18 +22,28 @@ class WhmcsConfig
     const SWITCH_ACCESS = 6;
     const DELETE_ACTION = 7;
     const PRE_INSTALL = 8;
+    const MEM_BILLING_ID = 9;
+    const DISK_BILLING_IDS = 10;
+    const ADDON_BILLING_IDS = 11;
+    const CLIENT_MANAGE_BUTTON = 12;
```
Comment (cabloo) [inline src/Whmcs/WhmcsConfig.php]: these should not be in this PR, they should be in a separate one
